### DISCUSSION TITLE
How to specify versions of gradle and jbake?

### DISCUSSION THREAD

**Discussion by lprimak** [docToolchain/docToolchain#1089] [Q&A] [answered]
I am having an issue with CI system having installed newer JDKs that docToolchain supports.
Also, I have installed jbake and gradle (latest).

The issue is that docToolchain uses older JBake and gradle that are not compatible with the latest JDK and don't run on Apple Silicon processor (where CI lives). Further complication is that jbake is an "RC" version that's not available in maven central yet.

Is there a way for docToolchain just use preinstalled gradle and jbake?

Thank you!

**Comment by rdmueller** [comment]
sorry that it took me so long to answer.

the problem with the jdk is solved through docToolchain having its own jdk (install through `./dtcw getJava` or by using the docker image directly in the build environment.

The reason why we are still on an older JDK is the asciidoctor gradle plugin used for `generateHTML` and `generatePDF`. This needs to be updated for newer gradle version. But the new gradle plugin has a completly different syntax.

Regarding the jBake version, I am sure there is a mechanism which allows us to drop the library somewhere in the file structure and then use it. But I havent tried it yet. Last time I took a look at it, it still mentioned that even the RC has problems with apple silicon. Is this resolved by now?

**Reply by lprimak** [reply]
> even the RC has problems with apple silicon

Yes, the latest (rc.7) jbake works with Apple Silicon. Latest JBake rc.7 is actually available via maven central as a library, along with the maven plugin.

Furthermore, the JBake (via maven plugin) can be upgraded to work with JDK 21 already (see https://github.com/flowlogix/website-template/blob/main/jbake-maven/pom.xml)

The issue that I am running into is integrating docToolchain into the current SSG system, which already runs on JBake among other things. Some things require newer versions of things, and I do not want to downgrade. I also don't want to have two different SSG runs (with possibly conflicting outputs and inputs).

I also noticed there is some asciidoc preprocessing going on outside JBake. Is this documented anywhere? What does that do?
Can I safely ignore it?

My other issue is gradle. All my other systems are maven and I want to desperately avoid gradle during the SSG generation on CI. Do you think that's at all possible?

**Comment by mh182** [comment, marked as answer]
We are currently stuck with Gradle 7.0.2 and Java 11 due to a bug in the Asciidoctor Gradle Plugin 3.* (for more details see #1102 ). As far as I can tell, JBake 2.7.0-rc.7 still has problems with Java 11 on Mac (ARM) - see [jbake release notes v2.7.0-rc.7](https://github.com/jbake-org/jbake/releases/tag/v2.7.0-rc.7).

As you can see from #1102 tracking the dependencies of the different components is not an easy task. This is why docToolchain vendors its components (this includes Gradle and the Java). We could not support all the different environments/versions.

**Reply by lprimak** [reply]
FYI: JBake works perfectly well with Java 20 on Apple Silicon, provided asm, groovy and javassist deps are updated.
With JBake maven plugin, this can be done transparently: https://github.com/flowlogix/website-template/blob/main/jbake-maven/pom.xml
